File: cms/app/audit.py
"""Audit log helper. Call log(...) from any write route to record what changed."""
import json
import logging

from fastapi import Request

from . import db


_log = logging.getLogger("piplayer.audit")


def log(
    request: Request | None,
    user: dict | None,
    action: str,
    target_type: str | None = None,
    target_id: str | int | None = None,
    details: dict | None = None,
) -> None:
    ip = None
    if request is not None and request.client is not None:
        ip = request.client.host
    try:
        with db.cursor() as cur:
            cur.execute(
                """INSERT INTO audit_log (user_id, username, action, target_type, target_id, details, ip)
                   VALUES (?, ?, ?, ?, ?, ?, ?)""",
                (
                    user["id"] if user else None,
                    user["username"] if user else None,
                    action,
                    target_type,
                    str(target_id) if target_id is not None else None,
                    json.dumps(details) if details else None,
                    ip,
                ),
            )
    except Exception:
        _log.exception("audit log write failed for %s", action)

File: cloud/src/audit.js
// Audit log helper. Call log(ctx, action, ...) from any write route to record what changed.
// The user is taken from ctx.user (pass `user` explicitly for login, where ctx.user is not
// set yet, or null for login_failed). Never throws: a failed audit write must not fail the request.
import * as db from "./db.js";
import { envInt } from "./util.js";

export function clientIp(ctx) {
  return ctx.request.headers.get("cf-connecting-ip") || null;
}

// json.dumps() with Python's default separators and ensure_ascii, so the Details column reads
// the same here as in the Python CMS: {"a": 1, "b": "x", "c": [1, 2]}.
export function pyJson(v) {
  if (v === null || v === undefined) return "null";
  if (Array.isArray(v)) return `[${v.map(pyJson).join(", ")}]`;
  if (typeof v === "object") {
    return `{${Object.entries(v).filter(([, x]) => x !== undefined).map(([k, x]) => `${pyJson(k)}: ${pyJson(x)}`).join(", ")}}`;
  }
  if (typeof v === "string") {
    return JSON.stringify(v).replace(/[\u0080-\uffff]/g, (c) => "\\u" + c.charCodeAt(0).toString(16).padStart(4, "0"));
  }
  return JSON.stringify(v);
}

export async function log(ctx, action, targetType = null, targetId = null, details = null, user = undefined) {
  const u = user === undefined ? ctx.user : user;
  try {
    await db.run(ctx.env,
      `INSERT INTO audit_log (user_id, username, action, target_type, target_id, details, ip)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      u ? u.id : null,
      u ? u.username : null,
      action,
      targetType,
      targetId === null || targetId === undefined ? null : String(targetId),
      details && Object.keys(details).length ? pyJson(details) : null,
      clientIp(ctx));
  } catch (e) {
    console.error(`audit log write failed for ${action}: ${e}`);
  }
}

// Daily: prune rows older than PIPLAYER_AUDIT_RETENTION_DAYS (0 = keep forever).
export function housekeeping(env) {
  return db.pruneAuditLog(env, envInt(env, "PIPLAYER_AUDIT_RETENTION_DAYS", 365));
}
